src/Development/IDE/Plugin/CodeAction.hs

```diff
@@ -770,15 +769,29 @@ suggestFunctionConstraint contents Diagnostic{..} missingConstraint
   | Just typeSignatureName <- findTypeSignatureName _message
   = let mExistingConstraints = findExistingConstraints _message
         newConstraint = buildNewConstraints missingConstraint mExistingConstraints
-        typeSignatureLine = findTypeSignatureLine contents typeSignatureName
-        typeSignatureFirstChar = T.length $ typeSignatureName <> " :: "
-        startOfConstraint = Position typeSignatureLine typeSignatureFirstChar
-        endOfConstraint = Position typeSignatureLine $
-          typeSignatureFirstChar + maybe 0 T.length mExistingConstraints
-        range = Range startOfConstraint endOfConstraint
-     in [(actionTitle missingConstraint typeSignatureName, [TextEdit range newConstraint])]
+     in case findRangeOfContextForFunctionNamed typeSignatureName of 
+       Just range -> [(actionTitle missingConstraint typeSignatureName, [TextEdit range newConstraint])]
+       Nothing -> []
   | otherwise = []
     where
+      findRangeOfContextForFunctionNamed :: T.Text -> Maybe Range 
+      findRangeOfContextForFunctionNamed typeSignatureName = do
+          locatedType <- listToMaybe 
+              [ locatedType
+              | L _ (SigD _ (TypeSig _ identifiers (HsWC _ (HsIB _ locatedType)))) <- hsmodDecls
+              , any (`isSameName` T.unpack typeSignatureName) $ fmap unLoc identifiers
+              ]
+          srcSpanToRange $ case locatedType of
+            -- The type signature has explicit Context
+            L _ (HsQualTy _ (L contextSrcSpan _ ) _) -> contextSrcSpan
```

Comment:
Not every type with a context is rooted in an `HsQualTy` constructor. Did you consider using `splitLHsQualTy` instead ?

Reply:
Good point, I wasn't aware a function like that existed.
I just tried it but it makes the test case where there's no initial context to start with fail.

Looking at how [splitLHsQualTy](https://hackage.haskell.org/package/ghc-8.10.1/docs/src/GHC.Hs.Types.html#splitLHsQualTy) is implemented it's clear why: It uses `noLHsContext = noLoc []` in the case where I use `SrcLoc` of the start of the type signature. So in that respect my "manual" pattern match seems better.

On the other hand `splitLHsQualTy` has an additional case `HsParTy` which I have no idea what kind of syntax it represents. I'll try to find out and tweak my implementation based on that (Ideally adding a test case which triggers that case).

Reply:
I didn't find an example syntax which would trigger that `HsParTy` case, but I reimplemented my attempt like this, which made the tests pass

```haskell
          srcSpanToRange $ case splitLHsQualTy locatedType of
            (L contextSrcSpan _ , _) ->
              if isGoodSrcSpan contextSrcSpan
                then contextSrcSpan -- The type signature has explicit context
                else -- No explicit context, return SrcSpan at the start of type sig where we can write context
                     let start = srcSpanStart $ getLoc locatedType in mkSrcSpan start start
```

BUT it seems to be noticably slower:

![Screenshot from 2020-10-04 16-14-14](https://user-images.githubusercontent.com/2716069/95017986-dc6fbb80-065c-11eb-807b-eab1f0fb2d8e.png)

compared to my previous impl:

![Screenshot from 2020-10-04 16-14-01](https://user-images.githubusercontent.com/2716069/95017992-ed203180-065c-11eb-81df-7dfdbfb2e367.png)

Which implementation would you suggest using?

Reply:
Nevermind, the speed difference seems to have been fluke. On repeated reruns both versions seems to perform the same. So I'm going with your suggestion of using the ghc helper function.